SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -176,7 +176,27 @@ void ComponentReader::ApplySchemaObject(Schema_Object* ComponentObject, UObject&
 		// Scoped to exclude OnRep callbacks which are already tracked per OnRep function
 		SCOPE_CYCLE_COUNTER(STAT_ReaderApplyPropertyUpdates);
 
-		for (uint32 FieldId : UpdatedIds)
+		// If we are applying initial data, they must have come from a ComponentData (as it currently stands).
+		// ComponentData will be missing fields if they are completely empty (options, lists, and maps).
+		// However, we still want to apply this empty data, so we need to reconstruct the full
+		// list of field IDs for that component type (Data, OwnerOnly).
+		const TArray<Schema_FieldId>& IdsToIterate = bIsInitialData ?
```

Comment:
We could construct this in ApplyComponentData instead of here and make this a bit simpler.
For updates we append cleared fields to the UpdateIds before passing it into here and it all works fine.

But for data we are constructing a list of update IDs then discarding it and creating a new one. then we are doing an expensive check to see that one is a subset of the other.
At the same time trying to get back the information that it is a data and not an update using a mislabelled variable :P.

Reply:
So, I did try to do this, the problem is that we require the RepLayout to reconstruct the field ids. We do not have it directly in ApplyComponentData, as we need the FObjectReplicator. We get that from `USpatialActorChannel::PreReceiveSpatialUpdate`, which I do not want to call twice or too early. The other option is to directly call
```
FindOrCreateReplicator(Object).Get()
```
But I am not sure about the side-effects of doing this. @m-samiec, any thoughts on creating the replicator early?

Reply:
Seems safe to call `FindOrCreateReplicator` as soon as you have the UObject and channel, so should be fine to call earlier.